Shanghai Update

pldespaigne opened the pull request:
Update the Yellow Paper from the _Paris_ hard-fork to the _Shanghai_ hard-fork.

- [x] EIP-3651: Warm COINBASE
- [x] EIP-3855: PUSH0 instruction
- [x] EIP-3860: Limit and meter initcode
- [x] EIP-4895: Beacon chain push withdrawals as operations
- [x] EIP-6049: Deprecate SELFDESTRUCT
- [x] General Updates

[Here](https://github.com/user-attachments/files/15835982/Paper.pdf) is the current pdf version of the Yellow Paper for this PR.


# EIP-3651: Warm COINBASE
- [specs](https://eips.ethereum.org/EIPS/eip-3651)
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/6fb1032933785381ad4859bfb4be0400e4c45fe5)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1606402480) → [change: 8af40f2](https://github.com/ethereum/yellowpaper/pull/906/commits/8af40f2457dab040ddb191139579335ec95a53df)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|![3651_before](https://github.com/ethereum/yellowpaper/assets/5858657/802da166-486c-4593-9ad1-b00583b8c81a)|![3651_after](https://github.com/ethereum/yellowpaper/assets/5858657/cbdafb35-afe2-400d-b291-723e2c01a8b4)|

</details>

# EIP-3855: PUSH0 instruction
- [specs](https://eips.ethereum.org/EIPS/eip-3855)
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/d997926306d953aaf12b8130ea2ce9b97c91e0a0)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1608345854)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|![3855_0_before](https://github.com/ethereum/yellowpaper/assets/5858657/8dbfc0b2-49d5-40e3-9afe-c5a80e9c63ab)|![3855_0_after](https://github.com/ethereum/yellowpaper/assets/5858657/990a6709-814c-4c3e-97ac-f2bc7d58e0aa)|
|![3855_1_before](https://github.com/ethereum/yellowpaper/assets/5858657/a7e14155-dd35-4b40-a88c-0fdd3a4d6dbd)|![3855_1_after](https://github.com/ethereum/yellowpaper/assets/5858657/d26c541d-b21c-4419-9164-73f60ba0b0c6)|

</details>

# EIP-3860: Limit and meter initcode
- [specs](https://eips.ethereum.org/EIPS/eip-3860)
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/8428ca552b587abed700aced1d2d96d9ff22104a)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1610052660) → [changes: 263171f](https://github.com/ethereum/yellowpaper/pull/906/commits/263171f3e9b7599475e9f0abc8efe4ed1b5040e9)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|![3860_0_before](https://github.com/ethereum/yellowpaper/assets/5858657/2f36d617-8b44-4f0f-9a77-9519c5857d2f)|![3860_0_after](https://github.com/ethereum/yellowpaper/assets/5858657/34b5b9a2-0074-4fb3-a6f2-35f9760033f1)|
|![3860_1_before](https://github.com/ethereum/yellowpaper/assets/5858657/5b0f6722-9c1d-4e59-b349-dc4c310140d4)|![3860_1_after_1](https://github.com/ethereum/yellowpaper/assets/5858657/1f8ac562-1478-4cb3-a18f-2194b76bc7ab)|
|![3860_2_before](https://github.com/ethereum/yellowpaper/assets/5858657/f0042b89-4c48-42c9-8e14-a93a5c593665)|![3860_2_after_1](https://github.com/ethereum/yellowpaper/assets/5858657/cb8a2723-b63b-4102-9766-693ca60c2e33)|
|![3860_3_before](https://github.com/ethereum/yellowpaper/assets/5858657/9e8bcda0-a3c6-4a05-a97e-955c2b49e613)|![3860_3_after_1](https://github.com/ethereum/yellowpaper/assets/5858657/0f52c0da-b9f1-4d9b-9a4d-8b0cc1923de2)|
|![3860_4_before](https://github.com/ethereum/yellowpaper/assets/5858657/340a3aae-56bb-4e3d-a22b-14056f39fb58)|![3860_4_after](https://github.com/ethereum/yellowpaper/assets/5858657/2bb81fbe-ce1c-498e-91af-3cbbc956d6a5)|

</details>

# EIP-4895: Beacon chain push withdrawals as operations
- [specs](https://eips.ethereum.org/EIPS/eip-4895)
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/5173cdd43e7185b5e62c6816120273fe53039e62)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1618729365) → [changes: aac8e1c](https://github.com/ethereum/yellowpaper/pull/906/commits/aac8e1c55acb03242b0d1e7123298362e087fba3)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|![4895_11_before](https://github.com/ethereum/yellowpaper/assets/5858657/acf7fbd2-d67c-41c6-9e28-0f03ca78d815)|![4895_11_after](https://github.com/ethereum/yellowpaper/assets/5858657/66aaa16a-474d-497d-abf6-c581ecc40f5d)|
|![4895_0_before](https://github.com/ethereum/yellowpaper/assets/5858657/14ce7e03-5aca-40c1-bbab-3cb463210935)|![4895_0_after](https://github.com/ethereum/yellowpaper/assets/5858657/518a9490-2377-4b94-92cb-d6a3455122b7)![4895_0_after_1](https://github.com/ethereum/yellowpaper/assets/5858657/354d76a5-6b32-4cba-a822-ab4f1d81b7b4)|
|![4895_1_before](https://github.com/ethereum/yellowpaper/assets/5858657/4cf3937d-3324-42fe-8429-8fe1e4668530)|![4895_1_after](https://github.com/ethereum/yellowpaper/assets/5858657/411d6e25-05dc-44cc-9a32-36f2634a82f5)|
|![4895_2_before](https://github.com/ethereum/yellowpaper/assets/5858657/140b659f-f53e-4c5c-9205-dd56a3a7ee95)|![4895_2_after](https://github.com/ethereum/yellowpaper/assets/5858657/b05dccd9-29c9-483c-97a2-4d1483aac62d)|
|![4895_3_before](https://github.com/ethereum/yellowpaper/assets/5858657/afbc9f26-5edc-4b52-b987-5f26f301580d)|![4895_3_after](https://github.com/ethereum/yellowpaper/assets/5858657/f4cfa4e6-fa06-4018-8e5b-9ab48a604da8)|
|![4895_4_before](https://github.com/ethereum/yellowpaper/assets/5858657/cd56daa6-80b0-4ce0-aef1-8c1a7850db23)|![4895_4_after](https://github.com/ethereum/yellowpaper/assets/5858657/d45970e7-918d-487d-b989-5778220a34ed)|
|![4895_5_before](https://github.com/ethereum/yellowpaper/assets/5858657/ac0c7351-ae4a-424e-b7d4-771571a19f5b)|![4895_5_after](https://github.com/ethereum/yellowpaper/assets/5858657/cea72bd0-0a51-407a-b806-0a17ca2bb6b4)|
|![4895_6_before](https://github.com/ethereum/yellowpaper/assets/5858657/e82279aa-0e84-4cef-8980-8623871f765f)|![4895_6_after](https://github.com/ethereum/yellowpaper/assets/5858657/02b969e3-fa6b-4c17-bc61-d921cc4af6c7)|
|![4895_7_before](https://github.com/ethereum/yellowpaper/assets/5858657/2b9901ce-342b-479f-90d4-f97d9954a1ea)|![4895_7_after](https://github.com/ethereum/yellowpaper/assets/5858657/42e85f1a-80ce-46d6-a3aa-36e87ae96fab)|
|![4895_8_before](https://github.com/ethereum/yellowpaper/assets/5858657/2b35e224-662b-4d2c-8fe1-45b2ab3d0b2f)|![4895_8_after](https://github.com/ethereum/yellowpaper/assets/5858657/a6e7a940-ce3f-45ba-923a-e00a5bdfabe5)|
|![4895_9_before](https://github.com/ethereum/yellowpaper/assets/5858657/a3b80746-2a3e-475f-8ccf-be7cb8b4e9b5)|![4895_9_after_1](https://github.com/ethereum/yellowpaper/assets/5858657/421d632e-2ff2-45f9-bbb1-7b09554f80fb)|
|![4895_10_before](https://github.com/ethereum/yellowpaper/assets/5858657/3483e508-5680-4d75-9058-8cc322b3a4ff)|![4895_10_after](https://github.com/ethereum/yellowpaper/assets/5858657/a5f781e9-5894-4e08-84d9-5af34149379f)|

</details>

# EIP-6049: Deprecate SELFDESTRUCT
- [specs](https://eips.ethereum.org/EIPS/eip-6049)
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/20ebaa7cb3947992b0c2a5bbae4881036f593372)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1618785791) → [changes: 22c2ca5](https://github.com/ethereum/yellowpaper/pull/906/commits/22c2ca5c9b56e180dfb5ae27b6b57b31daaa1ee2)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|![6049_0_before](https://github.com/ethereum/yellowpaper/assets/5858657/225e2e00-6139-4430-b6dc-f74d62b4301f)|![6049_0_after_1](https://github.com/ethereum/yellowpaper/assets/5858657/f258ae6d-4309-4cb8-b200-5a13a8ae5481)|

</details>

# General Updates
- [commit changes](https://github.com/ethereum/yellowpaper/pull/906/commits/5df35cf27feb0b7d5397db8f8592f0195868e93f)
- [review](https://github.com/ethereum/yellowpaper/pull/906#discussion_r1618795416) → [changes: 878ee5b](https://github.com/ethereum/yellowpaper/pull/906/commits/878ee5bd9977db043da8a293fe510737ee3be1d5)

<details>
<summary>View Screenshot Diff</summary>

|Before|After|
|-|-|
|![shanghai_0_before](https://github.com/ethereum/yellowpaper/assets/5858657/6c364d4e-82c3-418e-9f18-88146a4e8d87)|![shanghai_0_after](https://github.com/ethereum/yellowpaper/assets/5858657/a851d0ca-da21-46e3-b70f-9623eddd326d)|
|![shanghai_1_before](https://github.com/ethereum/yellowpaper/assets/5858657/54ca096a-2096-4c73-b14f-646a3a9712c8)|![shanghai_1_after](https://github.com/ethereum/yellowpaper/assets/5858657/4429fe0e-fb4a-4b75-a9c3-37c3ec2ed4cb)|
|![shanghai_2_before](https://github.com/ethereum/yellowpaper/assets/5858657/be4b5fe7-4e7f-44ef-b7fa-7f8689bf9ce2)|![shanghai_2_after](https://github.com/ethereum/yellowpaper/assets/5858657/43234e00-1aff-47a9-a61e-e7ad020bd090)|
|![shanghai_3_before](https://github.com/ethereum/yellowpaper/assets/5858657/65ebd220-cb57-422c-8bd1-5ecf0398fbc4)|![shanghai_3_after_1](https://github.com/ethereum/yellowpaper/assets/5858657/9a1eadde-552f-4171-8ce5-9ead14a87de7)|

</details>

pldespaigne commented:
Bump @yperbasis @timbeiko
Let me know, if you need more detail on the PR changes, or anything else!

yperbasis commented:
> Bump @yperbasis @timbeiko Let me know, if you need more detail on the PR changes, or anything else!

Currently I don't have the capacity to review this PR unfortunately :(

pldespaigne commented:
@yperbasis 
I understand, do you know of anyone that could have a look at this?
Should I try to bring some non-maintainer peoples in the convo to give feedback on my changes?
I've signed a grant agreement with the EF to update the YP to Shanghai & Cancun, that's why I'll try to help you as much as I can to merge this.

alexprut commented:
@pldespaigne not a maintainer, nor fully up to date with all the protocol changes, contributed (small) changes years ago.

However, the work you are doing is interesting, if it's helpful I can sped a spike next week to review and provide feedback.

pldespaigne commented:
That would be awesome 🙏
Anyone's feedback would be great!

alexprut commented on Paper.tex:
```diff
@@ -830,7 +830,7 @@ \subsection{Execution}
 a & \text{otherwise}
 \end{cases} \\
-a & \equiv A^0_{\mathbf{a}} \cup \{S(T)\} \cup_{E \in T_{\mathbf{A}}} \{ \hyperlink{access_list_entry}{E}_{\mathrm{a}} \}
+a & \equiv A^0_{\mathbf{a}} \cup \{S(T)\} \cup \{H_c\}\cup_{E \in T_{\mathbf{A}}} \{ \hyperlink{access_list_entry}{E}_{\mathrm{a}} \}
```
This part LGTM.

Small observations:
1) Are the curly brackets necessary? (i.e. can `\{H_c\}` just be  `H_c`). I guess because the union operation is defined on sets, if that is the criteria then formula `(74)` on `a ∪ T_t` is missing the brackets.
2) My understanding is that this EIP is retro-compatible, still should it specify in any comment/footnote that it was introduced with `EIP-3651`?
3) (_Out of scope of the PR_) While reading the Block definition section 4.3, `H_c` could also state the following: "(also known as the coinbase address)"

pldespaigne replied:
Re: EIP-3651
1. Looking at other formulas of the YP, you are right, I just removed `{ }` for clarity
2. I don't think so, EIPs are rarely cited except for major changes such as EIP1559.
3. Totally agree with you, I didn't added it though, cause it's out of scope and I don't want to add more noise, this PR will already be hard to merge, might be worth it to add it as a tiny separate PR.

pldespaigne replied:
Changes in [689eb33](https://github.com/ethereum/yellowpaper/pull/906/commits/689eb3353be76717d1378640b06551d73826d6a4)

alexprut commented on Paper.tex:
```diff
@@ -2115,9 +2195,9 @@ \subsection{Gas Cost}
+$W_{\mathrm{base}}$ = \{{\small ADDRESS}, {\small ORIGIN}, {\small CALLER}, {\small CALLVALUE}, {\small CALLDATASIZE}, {\small CODESIZE}, {\small GASPRICE}, {\small COINBASE},\newline \noindent\hspace*{1cm} {\small TIMESTAMP}, {\small NUMBER}, {\small PREVRANDAO}, {\small GASLIMIT}, {\small CHAINID}, {\small RETURNDATASIZE}, {\small POP}, {\small PC}, {\small MSIZE}, {\small GAS}, \newline \noind ...
 
-$W_{\mathrm{verylow}}$ = \{{\small ADD}, {\small SUB}, {\small NOT}, {\small LT}, {\small GT}, {\small SLT}, {\small SGT}, {\small EQ}, {\small ISZERO}, {\small AND}, {\small OR}, {\small XOR}, {\small BYTE}, {\small SHL}, {\small SHR}, {\small SAR}, \newline \noindent\hspace*{1cm} {\small CALLDATALOAD}, {\small MLOAD}, {\small MSTORE}, {\small MSTORE8}, {\small PUSH*}, {\small DUP*}, {\small SWA ...
+$W_{\mathrm{verylow}}$ = \{{\small ADD}, {\small SUB}, {\small NOT}, {\small LT}, {\small GT}, {\small SLT}, {\small SGT}, {\small EQ}, {\small ISZERO}, {\small AND}, {\small OR}, {\small XOR}, {\small BYTE}, {\small SHL}, {\small SHR}, {\small SAR}, \newline \noindent\hspace*{1cm} {\small CALLDATALOAD}, {\small MLOAD}, {\small MSTORE}, {\small MSTORE8}, {\small PUSH1}, ..., {\small PUSH32}, {\sm ...
```
Changes for `EIP-3855` LGTM.

- Not fully sure thought if formula nr `(157)` also needs a change (haven't spend much time on it)?

pldespaigne replied:
Not sure what formula you are talking about (the numbers changes almost with every commit).
Assuming you are talking about this formula:
![image](https://github.com/ethereum/yellowpaper/assets/5858657/8d7f3bb4-4f84-4eac-b20f-9e1de3a612b2)
If so, I don't think it need update as it only refer to $w  \in [PUSH1, PUSH32]$ and `PUSH0` is not in that range

alexprut replied:
Sounds great, indeed was referencing to the formula you posted.

alexprut commented on Paper.tex:
```diff
@@ -752,6 +752,12 @@ \subsection{Execution}
 $\hyperlink{G__accesslistaddress}{G_{\mathrm{accesslistaddress}}}$ and $\hyperlink{G__accessliststorage}{G_{\mathrm{accessliststorage}}}$ are the costs of warming up account and storage access, respectively.
 $G$ is fully defined in Appendix \ref{app:fees}.
 
+\hypertarget{initcode_cost}{}We define the \textit{initcode cost function}, formally $N$, as the amount of gas that needs to be paid for each word of the initcode prior to executing the creation code of a new contract:
```
Since this `EIP-3860` required a "network update", should it indicate that this was introduced with `EIP-3860`?

pldespaigne replied:
I'm not in favor as it's not a "major" change, but I'm open to discussion.

alexprut replied:
> I'm not in favor as it's not a "major" change, but I'm open to discussion.

Sounds good, don't have strong opinions on this one.

alexprut commented:
Soundness of `EIP-3860` changes LGTM.

pldespaigne commented on Paper.tex:
```diff
@@ -2030,6 +2045,8 @@ \section{Fee Schedule}\label{app:fees}
 \linkdest{G__selfdestruct}{}$G_{\mathrm{selfdestruct}}$ & 5000 & Amount of gas to pay for a {\small SELFDESTRUCT} operation. \\
 $G_{\mathrm{create}}$ & 32000 & Paid for a {\small CREATE} operation. \\
 \linkdest{G__codedeposit}{}$G_{\mathrm{codedeposit}}$ & 200 & Paid per byte for a {\small CREATE} operation to succeed in placing code into state. \\
+\linkdest{G__initcodeword}{}$G_{\mathrm{initcodeword}}$ & 2 & Paid per word of the initcode at the beginning of a deploy transaction, a {\small CREATE},\\
```
[EIP-3860](https://eips.ethereum.org/EIPS/eip-3860#parameters) define it as its own `INITCODE_WORD_COST` constant, also I would find it confusing to use $G_{base}$ because of the description: "Amount of gas to pay for operations of the set $W_{base}$.". What do you think?

In another subject, your comment made me realize that function $N(x)$ was already used as pointed in your review of `EIP-3855` above, I'll change it to $R(x)$ which I think is not used.

pldespaigne replied:
Changes in [#5a342a7](https://github.com/ethereum/yellowpaper/pull/906/commits/5a342a742d1c6c04b23b143a7c49b381b7a85660)

alexprut replied:
> [EIP-3860](https://eips.ethereum.org/EIPS/eip-3860#parameters) define it as its own `INITCODE_WORD_COST` constant, also I would find it confusing to use Gbase because of the description: "Amount of gas to pay for operations of the set Wbase.". What do you think?
> 
> In another subject, your comment made me realize that function N(x) was already used as pointed in your review of `EIP-3855` above, I'll change it to R(x) which I think is not used.

Your call, don't have strong opinions, it can always be simplified post this PR.

pldespaigne commented on Paper.tex:
```diff
@@ -2607,15 +2624,16 @@ \subsection{Instruction Set}
-&&&& is too low to fulfil the value transfer); and otherwise $x=\mathtt{ADDR}(I_{\mathrm{a}}, \boldsymbol{\sigma}[I_{\mathrm{a}}]_{\mathrm{n}}, \zeta, \mathbf{i} )$, the\\
-&&&& address of the newly created account (\ref{eq:new-address}). \\
+&&&& where $x=0$ if $z = 0$, i.e., the \hyperlink{contract_creation_result}{contract creation process failed},\\
+&&&& or $\lVert \mathbf{i} \rVert > 49152$ (initcode is longer than 49152 bytes),\\
```
You are right, EIP defines it as it's own constant, but since the YP used the same formalism on post deploy code check (Section 7), I decided to keep it like this.

![image](https://github.com/ethereum/yellowpaper/assets/5858657/99b6fbe8-619d-4da4-8206-2521ad65af54)

pldespaigne commented:
Thanks a lot for your feedbacks @alexprut 🙏

pldespaigne commented:
@nicksavers @pirapira Any time on your side to have a look at this?
Or maybe you could delegate that to someone?

alexprut commented:
> Thanks a lot for your feedbacks @alexprut 🙏

More than happy to be helpful, thanks for updating the YP.
@pldespaigne FYI I still need to look at and review commits `EIP-4895`, `EIP-6049` and `General Updates`, aiming at finishing the review by EOW.

alexprut commented on Paper.tex:
```diff
@@ -413,17 +413,45 @@ \subsection{The Transaction} \label{subsec:transaction}
+\item[globalIndex] zero based incrementing withdrawal index that acts as a unique identifier for this withdrawal; formally $W_{\mathrm{g}}$.
+\item[validatorIndex] index of consensus layer's validator this withdrawal corresponds to; formally $W_{\mathrm{v}}$.
+\item[recipient] the 20-bytes address that will receives Ether form this withdrawal; formally $W_{\mathrm{r}}$.
+\item[amount] a nonzero amount of Ether denominated in Gwei ($10^9$ Wei); formally $W_{\mathrm{a}}$.
```
`Gwei` could just go as a definition in section `2.1.`

pldespaigne replied:
Good catch! Since `EIP-4895` explicitly define this value as being denominated in `Gwei`, I think you are right.

pldespaigne replied:
changes in [55a4099](https://github.com/ethereum/yellowpaper/pull/906/commits/55a4099d6366a2425e4c7c8ad1ba2687cd1abc89)

alexprut commented on Paper.tex:
```diff
@@ -413,17 +413,45 @@ \subsection{The Transaction} \label{subsec:transaction}
 \end{equation}
 
+\subsection{The Withdrawal}\linkdest{withdrawal}\label{subsec:The_Withdrawal}
+A withdrawal (formally, \textit{W}) is a tuple of data describing a consensus layer validator's withdrawal of some amount of its staked Ether.
```
FYI, not a big issue, `W` is already defined in the paper as `W(w, μ)`

pldespaigne replied:
We don't have enough letters in greek & latin alphabet 😅
As you said, I think it's not a big issue, `W` for **W**ithdrawal is easy to remember, it follows the same logic as `T` for **T**ransaction, etc. Also, there is already `H` for block **H**eader even if the normal halting function is defined as `H(μ, I)`

alexprut commented on Paper.tex:
```diff
@@ -1365,10 +1404,27 @@ \section{Block Finalisation} \label{ch:finalisation}
 The process of finalising a block involves two stages:
```
3 stages now

pldespaigne replied:
changes in [55a4099](https://github.com/ethereum/yellowpaper/pull/906/commits/55a4099d6366a2425e4c7c8ad1ba2687cd1abc89)

alexprut commented on Paper.tex:
```diff
@@ -1365,10 +1404,27 @@ \section{Block Finalisation} \label{ch:finalisation}
 The process of finalising a block involves two stages:
 
 \begin{enumerate}
+\item executing withdrawals;
```
My understanding is that `validate transactions` precede  `executing withdrawals`

pldespaigne replied:
That's a good question! The EIP only states that it should happen after transaction *execution* but it says nothing about where it should happen during block finalization.
The transaction *validation* step purpose in the YP is simply meant to set the block header's gas used value ($B_{Hg}$), since withdrawals incur no gas cost, it has no impact if its run before or after withdrawal execution.

alexprut commented on Paper.tex:
```diff
@@ -1365,10 +1404,27 @@ \section{Block Finalisation} \label{ch:finalisation}
 
+\subsection{Executing Withdrawals}
+After processing the block's transactions, the withdrawals are executed. A withdrawal is simply an increase of the recipient account's balance of the specified amount. No other balances are decreased, a withdrawal is not a transfer but a creation of funds.
+A withdrawal operation cannot fail and has no gas cost. We define the function $P$ as the withdrawal state transition function:
```
FYI, not necessarily an issue, function `P` is already defined in the paper.

pldespaigne replied:
changes in [55a4099](https://github.com/ethereum/yellowpaper/pull/906/commits/55a4099d6366a2425e4c7c8ad1ba2687cd1abc89)

alexprut commented on Paper.tex:
```diff
@@ -2785,7 +2785,9 @@ \subsection{Instruction Set}
-\linkdest{selfdestruct}{}0xff & {\small SELFDESTRUCT} & 1 & 0 & Halt execution and register account for later deletion. \\
+\linkdest{selfdestruct}{}0xff & {\small SELFDESTRUCT} & 1 & 0 & Halt execution and register account for later deletion. Readers should be warned, \\
+&&&& since the \textit{Shanghai} update, this opcode has been marked as deprecated. \\
+&&&& Its behavior might be subject to change in an upcoming update. \\
```
Suggesting adding that it may be subject "to **breaking** changes"

pldespaigne replied:
changes in [91f0547](https://github.com/ethereum/yellowpaper/pull/906/commits/91f054787b4f38d15bddcefa09bc98ea0a62144b)

alexprut commented on Paper.tex:
```diff
@@ -1381,6 +1382,12 @@ \section{Transition to Proof of Stake} \label{ch:pos_transition}
+Because the Beacon Chain generate a new slot every 12 seconds, post-\textit{Paris} updates can be scheduled to occur at a specific timestamp.
+At the execution layer, the update will then happen in the first produced block after the scheduled timestamp.
+For example the \textit{Shanghai} hard fork was scheduled to occur at 2023-04-12 10:27:35 UTC, on Epoch 194,048.
+Validators failed to propose a block during the two first slots of this Epoch, but at slot 6,209,538 a validator finally proposed a the block 17,034,870, marking the transition to \textit{Shanghai} on the execution layer.
```
proposed **a the** block

pldespaigne replied:
changes in [2b98d68](https://github.com/ethereum/yellowpaper/pull/906/commits/2b98d68e5ea35e553fcc4fdab9c122ca91f55f38)

alexprut commented on Paper.tex:
```diff
@@ -203,6 +203,7 @@ \subsection{Which History?}
 $F_{\mathrm{Arrow Glacier}}$     & 13773000 \\
 $F_{\mathrm{Gray Glacier}}$      & 15050000 \\
 $F_{\mathrm{Paris}}$             & 15537394 \\
+$F_{\mathrm{Shanghai}}$          & 17034870 \\
```
Out of scope of this PR, was thinking if going forward there should be another column besides `Block Number`, e.g. `Timestamp` or whatever will be the canonical trigger.

pldespaigne replied:
That's a good idea imo!

yperbasis commented:
> Thanks again for your time @alexprut 🙏 I've made all the requested changes and update the PR description. I'll now try to bring more eyes on this discussion!
> 
> @yperbasis how many approvals would be required for you to be comfortable in merging this PR?

Sorry guys, I was on holiday and next week I'll be at a workshop. I'll take a look during the week starting from 24 June.

yperbasis reviewed (commented):
Please also update BRANCHES.md & README.md (see how it was done in PR #894 for Paris).

yperbasis commented:
Meanwhile I've created branch `paris` from the current `master`.

pldespaigne commented:
I've rebased the PR with latest changes from `master`.
I've also updated `README.md` & `BRANCHES.md` as requested.

Let me know if I can do anything else!

pldespaigne commented:
Thanks a lot for your time @yperbasis!
I'll now start to work on the Cancun update, should I open a single PR for the whole update, or do you prefer to review it eip by eip?

yperbasis commented:
> Thanks a lot for your time @yperbasis! I'll now start to work on the Cancun update, should I open a single PR for the whole update, or do you prefer to review it eip by eip?

Thank you! It's easier for me to review eip by eip.

pldespaigne commented:
Alright, I'll do it eip by eip then!
Also it looks like the merge of this PR didn't triggered a deploy of the new version of the pdf: https://ethereum.github.io/yellowpaper/paper.pdf

pldespaigne commented:
Thanks a lot 🙏
Will do, for sure!

yperbasis commented:
> Alright, I'll do it eip by eip then! Also it looks like the merge of this PR didn't triggered a deploy of the new version of the pdf: https://ethereum.github.io/yellowpaper/paper.pdf

I've re-run the job and https://ethereum.github.io/yellowpaper/paper.pdf seems to be fine now.
